### PR TITLE
do not expose mp_balance_mul, add MP_ENABLED feature detection

### PR DESCRIPTION
**Pull request by minad**
ping @czurnieden @sjaeckel

### CODE REVIEW

File: demo/test.c [hunk 1]
```diff
@@ -1,5 +1,8 @@
 #include "shared.h"
 
+/* We can also test the private API here */
+#include "tommath_private.h"
```
Comment (sjaeckel) [inline demo/test.c]: we can but should we?

IMO as soon as the library symbols are properly set-up you won't be able to build these tests then against the shared library

Comment (minad) [reply demo/test.c]: Then we need a special test build with exposed s_* symbols. I think it is better if internal functions are tested separately. For example in the mul case we could also test the different code paths by passing the right values. But this is too fragile and I would also like to test edge cases.

This would also solve the issue with exposed CUTOFF variables. We should just make those constant and test karatsuba and toom cook separately.

Comment (sjaeckel) [reply demo/test.c]: > We should just make those constant and test karatsuba and toom cook separately

IIUC they can't be made constant as they vary from platform to platform. Sure they're mostly constant now as nobody's really tweaking them but I guess that's caused by the fact that there's no proper way yet to determine them per platform. (besides running `etc/tune`)

> Then we need a special test build with exposed s_* symbols. I think it is better if internal functions are tested separately.

I'm no big fan of explicit testing of internal functionality.

> For example in the mul case we could also test the different code paths by passing the right values.

I prefer that way to make sure the entire chain works and then determine that everything is called through code coverage.

> But this is too fragile

why is this too fragile?

> I would also like to test edge cases

why would you want to test cases which can't be hit?

that all sounds a bit over the top to me

Comment (minad) [reply demo/test.c]: > I'm no big fan of explicit testing of internal functionality.

I am not saying to test arbitrary internal things. However internal functions with a well defined API should be tested, e.g., mp_balance_mul. @czurnieden also added the balance mul test to the test suite, but he did not call it directly which he should have. If only the external api should be tested, there should not even be a function test_balance_mul.

> I prefer that way to make sure the entire chain works and then determine that everything is called through code coverage.

If you want that - then the test suite must be split such that internal function tests and api function tests can be compiled separately. No big deal. `#if TEST_INTERNAL .... #endif`.
Ensure that stuff is hit via code coverage is good practice, but not enough and it depends on external tooling. Do we have coverage set up?

Besides, if we would just start renaming internal functions to `s_*`, we wouldn't have to tweak the visibility settings in the first step. The `s_*` functions would just be internal and not belong to the public ABI by our own convention. In the next step we could make the symbols hidden.

> why is this too fragile?

The test of the internal function mp_balance_mul should not depend on the criterion at which this function is selected in mp_mul.

>  why would you want to test cases which can't be hit?

Robustness of the functions. I would not extend functions to accept values which cannot be handled fundamentally, but for example if karatsuba also works for small numbers and these should be tested too.

> that all sounds a bit over the top to me

No it is not. If you want to see something which is over the top look at the sqlite test suite. What I am proposing here is good practice and modest. It is also too much work hopefully. Furthermore there are people willing to improve stuff, like me getting this test suite refactoring started.

**Comment by minad** [comment]
> TBH I preferred the way with separate if-statements with comments instead of one big if-statement with in-line comments

Hmm written like this it is more aligned with how the conditions are checked for the other special cases karatsuba and toom. I think it is was more clear like this, but astyle makes a bit of a mess :(

**Comment by minad** [comment]
@sjaeckel Better like this?

File: bn_mp_mul.c [hunk 2]
```diff
@@ -6,82 +6,69 @@
...
-   if ((MIN(len_a, len_b) < KARATSUBA_MUL_CUTOFF)
-       || ((MAX(len_a, len_b)) / 2 < KARATSUBA_MUL_CUTOFF)) {
-      goto GO_ON;
-   }
-   /*
-    * Not much effect was observed below a ratio of 1:2, but again: YMMV.
-    */
-   if ((MAX(len_a, len_b) /  MIN(len_a, len_b)) < 2) {
-      goto GO_ON;
+   if (0) {
```
Comment (sjaeckel) [inline bn_mp_mul.c]: srsly?

File: bn_mp_mul.c [hunk 3]
```diff
@@ -6,82 +6,69 @@
...
    }
 
-   res = mp_balance_mul(a,b,c);
-   goto END;
+#ifdef BN_MP_BALANCE_MUL_C
+   else if ((a->used != b->used) &&
+
+            /* Check sizes. The smaller one needs to be larger than the Karatsuba cut-off.
```
Comment (sjaeckel) [inline bn_mp_mul.c]: it's about the entire readability and such complex if statements suck

File: demo/test.c (hunk 1 above)
Comment (sjaeckel) [reply demo/test.c]: > If only the external api should be tested, there should not even be a function test_balance_mul.

why not if this test is specifically crafted to test `mp_balance_mul` within `mp_mul`? 

> Do we have coverage set up?

you can do `make coverage` :)

> Besides, if we would just start renaming internal functions to `s_*`, we wouldn't have to tweak the visibility settings in the first step. The `s_*` functions would just be internal and not belong to the public ABI by our own convention. In the next step we could make the symbols hidden.

fine by me, keep the discussion points of #172 in mind

> The test of the internal function mp_balance_mul should not depend on the criterion at which this function is selected in mp_mul.

what sense does a specific test even make if its counterpart isn't available? as it is now the test's can't even be built w/o `mp_balance_mul`

> if karatsuba also works for small numbers and these should be tested too.

then set the cutoff in the testcase appropriately?

File: bn_mp_mul.c (hunk 3 above)
Comment (minad) [reply bn_mp_mul.c]: Please compare the old code with the new code side by side, don't look at the diff only.
Then I think this PR improves things.

Comment (sjaeckel) [reply bn_mp_mul.c]: which old code? pre-206-merged?

Comment (minad) [reply bn_mp_mul.c]: Compare
https://github.com/libtom/libtommath/blob/develop/bn_mp_mul.c
with
https://github.com/libtom/libtommath/blob/41d3d73b8e19531fd476cc57355bdb1a015c5eb3/bn_mp_mul.c
and then tell me seriously if you defend the goto version.

Comment (sjaeckel) [reply bn_mp_mul.c]: Btw. I've the impression that this is going in the same direction as the rebase discussion.

Linearity of things, where linearity in my opinion leads to easier readability for a human.

Comment (minad) [reply bn_mp_mul.c]: I only see superficial similarities besides you disagreeing with me on basic things. And btw goto can be much less linear while if/else etc is linear by construction.

Comment (sjaeckel) [reply bn_mp_mul.c]: > besides you disagreeing with me on basic things

please explain

Comment (sjaeckel) [reply bn_mp_mul.c]: >  and then tell me seriously if you defend the goto version.

BTW yes, overall it improved

Comment (minad) [reply bn_mp_mul.c]: For example we disagreed about what should be tested or not, this is quite basic. Then there was the discussion about the rebasing, then the discussion about about the public ABI.

I would prefer if we wouldn't over discuss too many things to be honest. If there is something to be learned ok but otherwise it just costs time. It happened a few times now - I created a (from my pov) quick and uncontroversial patch and then a longish discussion emerges.  But I am not saying I am innocent in this respect ;)

Comment (minad) [reply bn_mp_mul.c]: The length of the patch should be proportional to the length of the discussion :)

Comment (minad) [reply bn_mp_mul.c]: > BTW yes, overall it improved

good to hear :sweat_smile:

Comment (sjaeckel) [reply bn_mp_mul.c]: > The length of the patch should be proportional to the length of the discussion :)

True, if there are no fundamentals being discussed as e.g. the testing.

> then the discussion about about the public ABI

You mean #172 ?

> I created a (from my pov) quick and uncontroversial patch

I was mostly fine with the original patch, besides the points I commented on. Then you made a v2 which implemented `mp_mul` entirely different and also changed parts that weren't commented on... yes this leads to re-discussion of things and new discussions...

**Comment by minad** [comment]
>     * I don't think `(a->used != b->used)` at line 18 will save enough to be worth the ops.

Fixed.

>     * And either use a variable for both MIN/MAX or for none. Please don't mix them. Many OCDers will thank you for that.

Ok, ok :)

> I can live with all of the above, no problem, with one exception: `if (0) {`

I will not change this one. I like it since it looks like a nice trick.

**Comment by minad** [comment]
Btw, we could introduce an `ENABLED` macro to handle the ifdefs then we could use something much nicer like. However we would then rely on the compilers dead code elimination. I usually prefer that style. If you want I prepare a patch, but this will change quite a few lines all over tommath I guess.

~~~
if (ENABLED(...) && condition) {
    ....
}
~~~

**Comment by czurnieden** [comment]
> . I like it since it looks like a nice trick.

I don't like it since it looks like an abomination.
An alien element.
A leftover you would automatically delete and then wonder why the compiler complains.

Please: either use a goto, or, since there is nothing to cleanup, set the sign and return from the balance branch directly.

Oh, and also: there is more to come. The TC4,5(6) can go below the balancing, no special treatment needed but FFT/NFT need a different kind of massage, must be treated before the balancing and have an immediate out, too. FFT/NFT will need some time (FFT is ready to go but NFT is not even started) but it would be nice if you don't make it too complicated to add them.

File: bn_mp_mul.c [hunk 4]
```diff
@@ -3,6 +3,13 @@
 /* LibTomMath, multiple-precision integer library -- Tom St Denis */
 /* SPDX-License-Identifier: Unlicense */
 
+#define _MP_ENABLED_TEST        ,
+#define _MP_ENABLED_TEST1       ,
+#define MP_ENABLED(x)           _MP_ENABLED1(x)
+#define _MP_ENABLED1(x)         _MP_ENABLED2(_MP_ENABLED_TEST##x)
+#define _MP_ENABLED2(x)         _MP_ENABLED3(x 1, 0)
+#define _MP_ENABLED3(x, y, ...) y
```
Comment (minad) [inline bn_mp_mul.c]: @czurnieden @sjaeckel How do you like something like this.

The macro will evaluate to 1 if the macro is either defined as empty or defined as 1.

Comment (czurnieden) [reply bn_mp_mul.c]: The original macros were meant to exclude code from compiling, from going into the binary in the first place. This patch keeps the branches in and hopes for the compiler to optimize them away.
Is it the case? Have you checked the dumps of GCC and clang? What is left if compiled with `-Os`?

Comment (minad) [reply bn_mp_mul.c]: This will always work. Every compiler optimizes `if (0)` away. For example the Linux kernel does that too. It is actually a quite common practice. The advantage of this is that you always get the typechecking.

Comment (minad) [reply bn_mp_mul.c]: The only situation where you have to go back to #ifdefs is if the definition is missing or changes depending on macros.

**Comment by minad** [comment]
superseded by #214
